torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+      for (auto tv : other_tv) {
+        tv->axis(0)->parallelize(ParallelType::BIDx);
+        if (rparams.num_warps > 1) {
+          tv->axis(2)->parallelize(ParallelType::TIDy);
+        }
+        tv->axis(-2)->parallelize(ParallelType::TIDx);
+      }
+
+      // Reduction TensorViews
+      for (auto tv : reduction_tv) {
+        tv->axis(0)->parallelize(ParallelType::BIDx);
+        if (rparams.num_warps > 1) {
+          tv->axis(2)->parallelize(ParallelType::TIDy);
+        }
+        tv->axis(-1)->parallelize(ParallelType::TIDx);
+      }
+
+      // rFactor TensorViews
+      for (auto tv : rfactor_tv) {
+        tv->axis(0)->parallelize(ParallelType::BIDx);
+        if (rparams.num_warps > 1) {
+          tv->axis(2)->parallelize(ParallelType::TIDy);
+        }
+        tv->axis(-2)->parallelize(ParallelType::TIDx);
+      }
+    }
+  } else {
+    const bool outer_axis_exists = reduction_tv.front()->nDims() > 2;
+    const int reduction_axis =
+        reduction_tv.front()->domain()->getReductionAxis().value();
+    const int inner_axis = reduction_axis - 1;
+    TORCH_INTERNAL_ASSERT(!outer_axis_exists || (inner_axis != 0));
+
+    // 2) Find TensorViews to duplicate and computeAt inline
+    auto duplicate_tv = findTensorViewsToDuplicate(fusion, other_tv);
+
+    // Any TVs with multiple uses and dependencies with same IterDomain
+    // Order of Duplication is necessary for correctness
+    for (auto tensor : duplicate_tv) {
+      auto result = tensor->duplicate();
+      // Add duplicated TVs to Other TVs
+      other_tv.insert(other_tv.end(), result.begin(), result.end());
+    }
+
+    // 3) For each reduction, apply reduction heuristics
+    std::vector<TensorView*> rfactor_tv;
+    for (auto tv : reduction_tv) {
+      bool rfactor_axis = false;
+
+      // Reduction Splits - [outer, inner, reduction-Leftover, TDX?]
+      if (rparams.lparams.bdimx() > 1) {
+        rfactor_axis = true;
+        tv->split(
+            reduction_axis, NamedScalar::getParallelDim(ParallelType::TIDx));
+      }
+
+      // Inner Splits
+      // [outer, inner-Leftover, BDY, TDY, reduction-Leftover, TDX?]
+      tv->split(inner_axis, NamedScalar::getParallelDim(ParallelType::TIDy));
+      tv->split(inner_axis, NamedScalar::getParallelDim(ParallelType::BIDy));
+
+      // Outer Splits
+      // [outer-lft, BDX?, inner-lft, BDY, TDY, reduction-lft, TDX?]
+      if (outer_axis_exists && rparams.lparams.gdimx() > 1) {
+        tv->split(0, NamedScalar::getParallelDim(ParallelType::BIDx));
+      }
+
+      if (rfactor_axis) {
+        auto reduction_tv_rf = tv->rFactor({-2});
+        rfactor_tv.push_back(reduction_tv_rf);
+      }
+    }
+
+    // 4) Other Tensor Splits
+    for (auto tv : other_tv) {
+      // Reduction Splits - [outer, inner, reduction-Leftover, TDX?]
+      if (rparams.lparams.bdimx() > 1) {
+        tv->split(
+            reduction_axis, NamedScalar::getParallelDim(ParallelType::TIDx));
+      }
```

Comment:
Isn't there a TV between the reduction and broadcast that will be missing a dimension? I'm missing the logic that handles that.

Reply:
Hmm, I don't understand. What TV are you thinking of?